modules/standard/Socket.chpl

```diff
@@ -0,0 +1,1403 @@
+  :return: connected socket.
+  :rtype: `tcpConn`
+  :throws SystemError: Upon failure to resolve address or connect
+                    to any of the resolved address in given `timeout`.
+*/
+proc connect(in host: string, in port: uint(16), family: IPFamily = IPFamily.IPUnspec, in timeout = new timeval(-1,0)): tcpConn throws {
+  return connect(host, port:string, family, timeout);
+}
+
+/*
+  A record holding reference to a udp socket
+  bound to any available port.
+*/
+record udpSocket {
+  /*
+    File Descriptor Associated with instance
+  */
+  var socketFd: int(32);
+  /* Create a UDP socket of provided Family. */
+  proc init(family: IPFamily = IPFamily.IPv4) {
+    this.socketFd = -1;
+    try! this.socketFd = socket(family, SOCK_DGRAM);
+    try! setBlocking(this.socketFd, false);
+  }
+
+  proc deinit() {
+    if this.socketFd != -1 {
+      sys_close(this.socketFd);
+    }
+  }
+}
+
+/* Get :type:`ipAddr` associated with udp socket */
+proc udpSocket.addr throws {
+  return getsockname(this.socketFd);
+}
+
+proc udpSocket.close throws {
+  var err_out = sys_close(this.socketFd);
+  if err_out != 0 {
+    throw SystemError.fromSyserr(err_out, "Failed to close udpSocket");
+  }
+}
+
+pragma "no doc"
+private extern proc sys_recvfrom(sockfd:fd_t, buff:c_void_ptr, len:size_t, flags:c_int, ref src_addr_out:sys_sockaddr_t, ref num_recvd_out:ssize_t):err_t;
+
+/*
+  Reads upto `bufferLen` bytes from the socket, and
+  return a tuple of (data, address), where address will be a
+  :type:`ipAddr` pointing to address of the socket from where data was received.
+
+  .. code-block:: Chapel
+
+    const timeout = new timeval(4,0);
+    const socket = new udpSocket();
+    const (data, sender) = socket.recvFrom(40, timeout);
+
+  :arg bufferLen: number of bytes to read
+  :type bufferLen: `int`
+  :arg timeout: time to wait for data to arrive.
+  :type timeval: :type:`~Sys.timeval`
+  :return: tuple of (data, address)
+  :rtype: `tuple(:mod:`bytes <Bytes>`, :type:ipAddr)`
+  :throws SystemError: Upon failure to receive any data
+                    within given `timeout`.
+*/
+proc udpSocket.recvfrom(bufferLen: int, in timeout = new timeval(-1,0), flags:c_int = 0):(bytes, ipAddr) throws {
+  var err_out:err_t = 0;
+  var buffer = c_calloc(c_uchar, bufferLen);
+  var length:ssize_t;
+  var addressStorage = new sys_sockaddr_t();
+  err_out = sys_recvfrom(this.socketFd, buffer, bufferLen:size_t, 0, addressStorage, length);
+  if err_out == 0 {
+    return (createBytesWithOwnedBuffer(buffer, length, bufferLen), new ipAddr(addressStorage));
+  }
+  if err_out != 0 && err_out != EAGAIN && err_out != EWOULDBLOCK {
+    c_free(buffer);
+    throw SystemError.fromSyserr(err_out,"recv failed");
+  }
+  var localSync$: sync c_short;
+  var internalEvent = event_new(event_loop_base, this.socketFd, EV_READ | EV_TIMEOUT, c_ptrTo(syncRWTCallback), c_ptrTo(localSync$):c_void_ptr);
+  defer {
+    event_free(internalEvent);
+  }
+
+  while err_out != 0 {
```

Comment:
same question as above, should we be worried about yielding here?

Reply:
The function call in `readFE()` which yields the thread and that will yield and only return when the running event loop passes down an event to the `sync` variable or a timeout occurs which is also passed down by the event loop to `sync` variable.